Blockchain gas support

AzhanL opened the pull request:
Adding function execution for the blockchain

AzhanL committed:
Update blockchain.py

Execute functions that need to be executed for contract accounts. Function names and args are retrieved from the data field of the transaction

AzhanL committed:
Update blockchain.py

Erase old gas calculation

AzhanL committed:
Update blockchain.py

Type hinting mistake

Lobarr commented on lib/blockchain.py:
```diff
@@ -174,14 +174,20 @@ def choose_transactions_from_mempool(self) -> List[Transaction]:
-          chosen_transactions.append(chosen_transaction)
+      chosen_transaction: Transaction = self.choose_random_transaction(transaction_bucket)
+      sender_account: Account, recipient_account: Account = self.get_transaction_accounts(chosen_transaction)
+      if sender_account.is_contract():
```
Why sender_account? I think it should be recipient_account so any account can execute the contract

AzhanL replied:
Because  the sender would not be transacting to anyone but rather just executing change state?

Lobarr replied:
Right but it's basically a transaction between the sender and the contract. Plus sender gets charged for executing the contract.

AzhanL replied:
Then who would the recipient be? The contract address?

Lobarr replied:
Yeah the contract address. That's why I implemented it the way I did.

Lobarr commented on lib/blockchain.py:
```diff
@@ -176,12 +176,18 @@ def choose_transactions_from_mempool(self) -> List[Transaction]:
+        # Retreive the function arguments from the data
+        func_args = chosen_transaction.get_data()['func_args']
+        # Check if they are the correct type
+        if type(func_name) is str and type(func_args) is tuple:
```
btw why do u have to check the type again if you're setting them from the opcode?

AzhanL replied:
So i'm not setting it from opcode. I'm using the name of the function to execute the function. Let me fix up the account. It'll be more clear then

AzhanL replied:
So here it is checking the data field of the transaction to check if they were properly sent in case it was malformed. Just an extra precaution

AzhanL committed:
Update blockchain.py

Send -> recipient

AzhanL committed:
Updated account.py

Add gas support

AzhanL committed:
Delete account.py

Wrong folder

AzhanL committed:
Update account.py

Added: Gas support + Gas calculation
get_body: has been modified to remove actual functions during export

Lobarr commented on lib/account.py:
```diff
@@ -25,9 +40,24 @@ def set_balance(self, balance: float):
+    if security_clearance:
+      self.contract_hash = Helper.hash_data(self.contract_code.encode())
+
+    if security_clearance:
```
This statement can be merged with the above one & you can remove the print
```suggestion
    if security_clearance:
    	self.contract_hash = Helper.hash_data(self.contract_code.encode())
    	self.generateFunctions(self.contract_code)
    return security_clearance
```

AzhanL committed:
Update account.py

combine security clearance

Lobarr commented on lib/account.py:
```diff
@@ -50,13 +79,36 @@ def withdraw(self, amount: int) -> bool:
-  def charge_gas(self, gas: float) -> bool:
+    return 0
+
+  def charge_gas(self, gas: float, func_name: str, func_args: tuple) -> bool:
```
How are you supposed to charge an external account if it executes a contract? This implementation seems like you're charging a contract for being run.

AzhanL committed:
Update blockchain.py

Charge the external account.

AzhanL committed:
Fix who's being charged

Doesn't charge the contract. Simply returns if it has been charged or not

Lobarr commented on lib/account.py:
```diff
@@ -89,5 +147,93 @@ def make_account(account_data: dict):
+    # Initializes the contract with the functions
+    injected_context = {
+        'state': {},
+        'data': {}
```
Include the msg object 
```
{
	'msg': {
		'sender': 'some-sender'
	}
}
```

AzhanL committed:
Update helper.py

update generate_address to use os.urandom for more randomness

Lobarr committed:
fix: minor cleanup

Lobarr commented on lib/helper.py:
```diff
@@ -1,7 +1,10 @@
 import hashlib
 import json
+import os
 from uuid import uuid4
```
feel free to remove this import since it's not being used anymore

AzhanL committed:
Update blockchain.py

Gas calculations correction

AzhanL committed:
Update account.py

Add msg for injected context and fix up state

AzhanL committed:
Create demo-script.py

Demo script for testing

AzhanL committed:
Update sample.py

Sample contrac

AzhanL committed:
Update sample.py

Added admin checks

Lobarr committed:
progress

Lobarr committed:
Merge branch 'blockchain-gas-support' of github.com:Lobarr/peth into fix/minor-changes

AzhanL committed:
Update app.py

Edit create_transaction return to be more description

Lobarr committed:
progress

Lobarr committed:
progress

Lobarr committed:
Merge branch 'master' of github.com:Lobarr/peth into fix/minor-changes

Lobarr committed:
Fix/minor changes (#10)

Fix/minor changes

Lobarr committed:
removed dup GAS_PRICE
